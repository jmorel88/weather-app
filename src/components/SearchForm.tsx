import { useState } from "react";

interface Props {
  getWeatherByZipCode: (zipCode: string) => void;
  getWeatherByGeolocation: () => void;
  error: Error | null;
  loading: boolean;
}

function SearchIcon() {
  return (
    <svg
      className="w-4 h-4"
      aria-hidden="true"
      xmlns="http://www.w3.org/2000/svg"
      fill="none"
      viewBox="0 0 20 20"
    >
      <path
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
      />
    </svg>
  );
}

export default function SearchForm({
  getWeatherByZipCode,
  getWeatherByGeolocation,
  loading,
  error,
}: Props) {
  const [zipCode, setZipCode] = useState<string>("");

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    if (loading) return;
    getWeatherByZipCode(zipCode);
  };

  const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    setZipCode(e.target.value);
  };

  return (
    <form className="w-full max-w-5xl mx-auto" onSubmit={handleSubmit}>
      {error && <p className="text-red-700 text-xs mb-2">{error.message}</p>}

      <label className="text-white-800 text-xs mb-2 block" htmlFor="zipCode">
        Zip Code
      </label>

      <div className="relative mb-2">
        <input
          className="bg-black-200 text-white-800 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full ps-10 p-3"
          id="zipCode"
          onChange={handleOnChange}
          value={zipCode}
          name="zipCode"
          placeholder="Search city by zip code"
          type="text"
        />

        <button
          className="text-white-800 absolute h-full aspect-square flex items-center justify-center top-0 disabled:opacity-50"
          type="submit"
          disabled={loading}
          aria-label="Submit zip code"
        >
          <SearchIcon />
        </button>
      </div>

      <div className="text-center md:flex md:justify-between w-full">
        <small className="text-xs text-gray-600 block">
          Format should be 5 digits (e.g. 10001)
        </small>
        <button
          className="text-xs text-blue-500 underline disabled:opacity-50"
          type="button"
          disabled={loading}
          onClick={getWeatherByGeolocation}
        >
          Search current location
        </button>
      </div>
    </form>
  );
}
