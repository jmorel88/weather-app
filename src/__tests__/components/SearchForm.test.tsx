import { afterEach, describe, test, vi } from "vitest";
import { fireEvent, render, screen } from "@testing-library/react";
import SearchForm from "../../components/SearchForm";

const getWeatherByZipCodeMock = vi.fn();
const getWeatherByGeolocationMock = vi.fn();

describe("SearchForm", () => {
  afterEach(() => {
    vi.resetAllMocks();
  });

  test("renders default correctly", ({ expect }) => {
    const { container } = render(
      <SearchForm
        error={null}
        loading={false}
        getWeatherByZipCode={getWeatherByZipCodeMock}
        getWeatherByGeolocation={getWeatherByGeolocationMock}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test("on search, call getWeatherByZipCode", ({ expect }) => {
    render(
      <SearchForm
        error={null}
        loading={false}
        getWeatherByZipCode={getWeatherByZipCodeMock}
        getWeatherByGeolocation={getWeatherByGeolocationMock}
      />,
    );

    const input = screen.getByLabelText("Zip Code");
    const button = screen.getByRole("button", { name: "Submit zip code" });

    fireEvent.change(input, { target: { value: "10001" } });
    fireEvent.click(button);

    expect(getWeatherByZipCodeMock).toHaveBeenCalledWith("10001");
  });

  test("search by current location calls getWeatherByGeolocation", ({
    expect,
  }) => {
    render(
      <SearchForm
        error={null}
        loading={false}
        getWeatherByZipCode={getWeatherByZipCodeMock}
        getWeatherByGeolocation={getWeatherByGeolocationMock}
      />,
    );
    const button = screen.getByRole("button", { name: /search current location/i });

    fireEvent.click(button);

    expect(getWeatherByGeolocationMock).toHaveBeenCalled();
  });

  test("on error, display error message", ({ expect }) => {
    render(
      <SearchForm
        error={new Error("Error message")}
        loading={false}
        getWeatherByZipCode={getWeatherByZipCodeMock}
        getWeatherByGeolocation={getWeatherByGeolocationMock}
      />,
    );

    expect(screen.getByText("Error message")).toBeInTheDocument();
  });

  test("on loading, buttons are disabled", ({ expect }) => {
    render(
      <SearchForm
        error={null}
        loading={true}
        getWeatherByZipCode={getWeatherByZipCodeMock}
        getWeatherByGeolocation={getWeatherByGeolocationMock}
      />,
    );

    const zipcodeButton = screen.getByRole("button", { name: "Submit zip code" });
    const geolocationButton = screen.getByRole("button", { name: /search current location/i });

    expect(zipcodeButton).toBeDisabled();
    expect(geolocationButton).toBeDisabled();

  });
});
