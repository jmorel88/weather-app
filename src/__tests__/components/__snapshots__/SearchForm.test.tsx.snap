// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SearchForm > renders default correctly 1`] = `
<div>
  <form
    class="w-full max-w-5xl mx-auto"
  >
    <label
      class="text-white-800 text-xs mb-2 block"
      for="zipCode"
    >
      Zip Code
    </label>
    <div
      class="relative mb-2"
    >
      <input
        class="bg-black-200 text-white-800 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full ps-10 p-3"
        id="zipCode"
        name="zipCode"
        placeholder="Search city by zip code"
        type="text"
        value=""
      />
      <button
        aria-label="Submit zip code"
        class="text-white-800 absolute h-full aspect-square flex items-center justify-center top-0 disabled:opacity-50"
        type="submit"
      >
        <svg
          aria-hidden="true"
          class="w-4 h-4"
          fill="none"
          viewBox="0 0 20 20"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
    </div>
    <div
      class="text-center md:flex md:justify-between w-full"
    >
      <small
        class="text-xs text-gray-600 block"
      >
        Format should be 5 digits (e.g. 10001)
      </small>
      <button
        class="text-xs text-blue-500 underline disabled:opacity-50"
        type="button"
      >
        Search current location
      </button>
    </div>
  </form>
</div>
`;
